### PR TITLE
Long khuong/api for registration

### PR DESCRIPTION
**Pull request by khuolo01**
- Create a simple sign-in and sign-up template for testing purposes.
- Create test files for user_sign_in and user_sign_up.
- API endpoint for sign-in and sign-up.
- Populate test database with few data.
- Modify conflicts in users' ids in test_user after populating test data.

### CODE REVIEW

**Comment by leth02** [comment]
For unit testing, this is not how you do it.

When we test the routes, we are not actually running any Flask server and send requests to that server.

What we do is actually, in the test file, say that we will override the output of a certain function (aka route) with a piece of sample data.

Then in the test, we call the function with certain input and expect that given that input, the function will return the output above.

So, to sum up
- we do not run any real server with unit testing.
- we override the output of the function, or even any library function call (e.g requests.post) with a sample output.

Please let me know here if things are not clear to you.

File: config.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import os
+
+# Safely transfer data
+
+KEY= os.environ.get("SECRET_KEY") or b'\xb0\x9a\xc9\x05\x19\xafv\x88\xfaB\xe5$\x07\x12\x10\x9e'
```
Comment (leth02) [inline config.py]: What is the use of this KEY? I see that you import it in the __init__.py but don't use it at all. I saw that you use it in the test. Why did not you use it in the app itself?

File: tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import sys
+sys.path.insert(0, 'D:\Luther\Self-Learning\Chat App\VL_Chat_App')
```
Comment (leth02) [inline tests/conftest.py]: You cannot add a local path from your computer to the codebase like this...

**Comment by leth02** [comment]
Do you think we should use an ORM to execute database requests?

File: config.py (hunk 1 above)
Comment (khuolo01) [reply config.py]: Sorry for the confusion. This commit is just a demo of how to push code to git I showed Phat earlier. I haven't finished fixing the test yet :D

File: tests/conftest.py (hunk 2 above)
Comment (khuolo01) [reply tests/conftest.py]: Same as above, I haven't finished fixing the test yet. This commit is just a demo.

**Comment by khuolo01** [comment]
> Do you think we should use an ORM to execute database requests?

Yes, we should. I'm actually working on it. It will make the code a lot easier to read.

File: db/db.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import sqlite3
+import os
+from flask import g, current_app
+
+# Sources:
```
Comment (leth02) [inline db/db.py]: What kind of sources? Why do you include it here?

File: db/db.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+# https://flask.palletsprojects.com/en/2.0.x/patterns/sqlite3/
+# https://flask.palletsprojects.com/en/2.0.x/tutorial/database/
+
+DATABASE = os.path.join("db", "VL_MESSAGES.db")
```
Comment (leth02) [inline db/db.py]: What does this variable do?

File: db/db.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+# Connect to the database
+def get_db():
+    db = getattr(g, '_database', None)
```
Comment (leth02) [inline db/db.py]: Add a comment to explain what g and _database do?

File: db/db.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    if db is None:
+        db = g._database = sqlite3.connect(DATABASE)
+        db.row_factory = sqlite3.Row
+    return db
+
+# Close the connection
```
Comment (leth02) [inline db/db.py]: What kind of connection this is?

File: db/db.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+def close_db(e=None):
+    db = g.pop('db', None)
+
+    if db is not None:
+        db.close()
+
+# Convert the selected rows into dictionary
```
Comment (leth02) [inline db/db.py]: Add comment?

File: db/db.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+def make_dicts(cursor, row):
+    return dict((cursor.description[idx][0], value)
+                for idx, value in enumerate(row))
+
+# Run a query
```
Comment (leth02) [inline db/db.py]: Please add a more descriptive comment.

File: db/db.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+def query_db(query, args=(), one=False):
+    cur = get_db().execute(query, args)
+    rv = cur.fetchall()
+    cur.close()
+    return (rv[0] if rv else None) if one else rv
+
+# Init db if data doesn't exist
```
Comment (leth02) [inline db/db.py]: Need better explanation here. What is _data_?

File: db/schema.sql [hunk 10]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE IF NOT EXISTS users (
```
Comment (leth02) [inline db/schema.sql]: As this script will be used to set up a new development database, we should have a `DROP TABLE IF EXIST` statement for each table.

File: db/schema.sql [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+    username text NOT NULL PRIMARY KEY,
```
Comment (leth02) [inline db/schema.sql]: **Primary key should be ID**, not username

File: message_app/__init__.py [hunk 12]
```diff
@@ -1,73 +1,140 @@
+import sys
+import os
+sys.path.append(os.getcwd())
```
Comment (leth02) [inline message_app/__init__.py]: If this line is to help run the test, we should have it in the test repo, somewhere in the test file?

File: message_app/__init__.py [hunk 13]
```diff
@@ -1,73 +1,140 @@
...
-        except Exception as error:
-            return {"Error": "Bad request. " + str(error)}, 400
-
-    return app
\ No newline at end of file
+def create_app(test_config=None):
+	app = Flask(__name__)
+
+	app.config.from_mapping(
+		SECRET_KEY=b'\xb0\x9a\xc9\x05\x19\xafv\x88\xfaB\xe5$\x07\x12\x10\x9e'
```
Comment (leth02) [inline message_app/__init__.py]: What is SECRET_KEY? What is it used for? Should we have a hex string SECRET_KEY?

**Comment by leth02** [comment]
Use snake case for variable naming(e.g. this_is_something = "something")
